# <span style = 'color:Gold'> <center> CREDIT CARD FRAUD DETECTION USING MACHINE LEARNING </center> </span>
---

## <span style = 'color:Indigo'> IMPORTING NECESSARY LIBRARIES </span>
---

In [1]:
import warnings
import numpy as np
import pandas as pd

warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

----

## <span style = 'color:Indigo'> DATA EXTRACTION </span>
---

In [2]:
creditCard_df = pd.read_csv('Dataset/creditcard.csv')

---

## <span style = 'color:Indigo'> EXPLORATORY DATA ANALYSIS </span>
---

### <span style = 'color:Green'> <i> Head of the dataset (Initial rows of the dataframe) </i> </span>
---

In [3]:
creditCard_df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9       V10       V11       V12       V13       V14  \
0  0.098698  0.363787  0.090794 -0.551600 -0.617801 -0.991390 -0.311169   
1  0.085102 -0.255425 -0.166974  1.612727  1.065235  0.489095 -0.143772   
2  0.247676 -1.514654  0.207643  0.624501  0.066084  0.717293 -0.165946   
3  0.377436 -1.387024 -0.054952 -0.226487  0.178228  0.507757 -0.287924   
4 -0.270533  0.817739  0.753074 -0.822843  0.538196  1.345852 -1.119670   

        V15       V16       V17       V18       V19       V20       V21  \
0  1.468177 -0.470401  0.207971  0.025791  0.403993  0.251412 -0.018307   
1  0.635558  0.463917 -0.114805 -0.183361 -0.145783 -0.069083 -0.225775   
2  2.345865 -2.890083  1.109969 -0.121359 -2.261857  0.524980  0.247998   
3 -0.631418 -1.059647 -0.684093  1.965775 -1.232622 -0.208038 -0.108300   
4  0.175121 -0.451449 -0.237033 -0.038195  0.803487  0.408542 -0.009431   

        V22       V23       V24       V25       V26       V27       V28  \
0  0.277838 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053   
1 -0.638672  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724   
2  0.771679  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752   
3  0.005274 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458   
4  0.798278 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   

   Amount  Class  
0  149.62      0  
1    2.69      0  
2  378.66      0  
3  123.50      0  
4   69.99      0

### <span style = 'color:Green'> <i> Tail of the dataset (Final rows of the dataframe) </i> </span>
---

In [4]:
creditCard_df.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9       V10       V11       V12  \
284802 -2.606837 -4.918215  7.305334  1.914428  4.356170 -1.593105  2.711941   
284803  1.058415  0.024330  0.294869  0.584800 -0.975926 -0.150189  0.915802   
284804  3.031260 -0.296827  0.708417  0.432454 -0.484782  0.411614  0.063119   
284805  0.623708 -0.686180  0.679145  0.392087 -0.399126 -1.933849 -0.962886   
284806 -0.649617  1.577006 -0.414650  0.486180 -0.915427 -1.040458 -0.031513   

             V13       V14       V15       V16       V17       V18       V19  \
284802 -0.689256  4.626942 -0.924459  1.107641  1.991691  0.510632 -0.682920   
284803  1.214756 -0.675143  1.164931 -0.711757 -0.025693 -1.221179 -1.545556   
284804 -0.183699 -0.510602  1.329284  0.140716  0.313502  0.395652 -0.577252   
284805 -1.042082  0.449624  1.962563 -0.608577  0.509928  1.113981  2.897849   
284806 -0.188093 -0.084316  0.041333 -0.302620 -0.660377  0.167430 -0.256117   

             V20       V21       V22       V23       V24       V25       V26  \
284802  1.475829  0.213454  0.111864  1.014480 -0.509348  1.436807  0.250034   
284803  0.059616  0.214205  0.924384  0.012463 -1.016226 -0.606624 -0.395255   
284804  0.001396  0.232045  0.578229 -0.037501  0.640134  0.265745 -0.087371   
284805  0.127434  0.265245  0.800049 -0.163298  0.123205 -0.569159  0.546668   
284806  0.382948  0.261057  0.643078  0.376777  0.008797 -0.473649 -0.818267   

             V27       V28  Amount  Class  
284802  0.943651  0.823731    0.77      0  
284803  0.068472 -0.053527   24.79      0  
284804  0.004455 -0.026561   67.88      0  
284805  0.108821  0.104533   10.00      0  
284806 -0.002415  0.013649  217.00      0

### <span style = 'color:Green'> Information about the dataset </span>
---

In [5]:
creditCard_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

### <span style = 'color:Green'> <i> Checking for Missing Values in the Dataframe </i> </span>
---

   * **There are no missing records in the dataset.**

In [6]:
round((creditCard_df.isnull().sum() / len(creditCard_df)) * 100, 2)

Time      0.0
V1        0.0
V2        0.0
V3        0.0
V4        0.0
V5        0.0
V6        0.0
V7        0.0
V8        0.0
V9        0.0
V10       0.0
V11       0.0
V12       0.0
V13       0.0
V14       0.0
V15       0.0
V16       0.0
V17       0.0
V18       0.0
V19       0.0
V20       0.0
V21       0.0
V22       0.0
V23       0.0
V24       0.0
V25       0.0
V26       0.0
V27       0.0
V28       0.0
Amount    0.0
Class     0.0
dtype: float64

### <span style = 'color: Green'> <i> Shape of the Dataset </i> </span>
---
    
   * The dataset contains <span style = 'color: Red'>**284807**</span> observations (records) and <span style = 'color: Red'>**31**</span> features

In [7]:
creditCard_df.shape

(284807, 31)

### <span style = 'color: Green'> <i> Number of Legitimate and Fradulant Transactions </i> </span>
---

In [8]:
creditCard_df.Class.value_counts()

0    284315
1       492
Name: Class, dtype: int64

#### <span style = 'color: Orange'> <i> Precentage of Legitimate and Fradulant Transactions </i> </span>
---

In [9]:
round((creditCard_df['Class'].value_counts() / len(creditCard_df)) * 100, 2)

0    99.83
1     0.17
Name: Class, dtype: float64

#### <span style = 'color:Orange'> <i> Observations </i> </span>
---

   * The **'Class'** Feature is classified into 2 categories:
   
   
       ➡️ 0: Legitimate Transactions
       ➡️ 1: Fraudulent Transactions
       
   * The dataset it highly imbalanced.
   
   
   * From the above block of code, it can seen that the dataset contains <span style = 'color:Red'>**99.83%**</span> of <span style = 'color:Red'>**legit transactions**</span> where as only <span style = 'color:Red'> **0.17%** </span> is <span style = 'color:Red'>**fraudulant transactions.**</span>
   
   
   * More than 90% of the transactions have been classified as Legitimate due which the data is highly imbalanced.
   
   
   * This dataset cannot be fed to the Predictive model. If fed, then if any fraudulent transaction comes as input, the model will classify it as legitimate.

---

## <span style = 'color:Indigo'> DATA PRE-PROCESSING </span>
---

### <span style = 'color:Green'> <i> Seperate the Class Feature for further Analysis </i> </span>
---

  * All the transactions which are legitimate will be stored in variable called **"legitimate_Transactions"**.
  

  * All the transactions which are fradulant will be stored in variable called ***"Fradulent_Transactions"**.

In [10]:
legitimate_Transactions = creditCard_df[creditCard_df.Class == 0]
Fraudulent_Transactions = creditCard_df[creditCard_df.Class == 1]

### <span style = 'color:Green'> <i> Shape of Legitimate and Fraudulent Transactions </i> </span>
---

   * Once the Class is separated
       
       ➡️ the Legitimate transactions have **284315** records and **3** Features.
       
       ➡️ the Fraudulent transactions have **492** records and **31** Features.

In [11]:
print(f"The shape of the Legitimate Transactions : {legitimate_Transactions.shape}")
print()
print(f"The shape of the Fradulent Transactions : {Fraudulent_Transactions.shape}")

The shape of the Legitimate Transactions : (284315, 31)

The shape of the Fradulent Transactions : (492, 31)


### <span style = 'color:Green'> <i> Statistical Measures of the Dataframe </i> </span>
---

#### <span style = 'color:Orange'> Legitimate Transactions </span>
---

In [12]:
print(f"Decription of the Amount feature of Legit transactions :")
legitimate_Transactions.Amount.describe()

Decription of the Amount feature of Legit transactions :


count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

#### <span style = 'color:Orange'> Fraudulent Transactions </span>
---

In [13]:
print(f"Decription of the Amount feature of Fraud transactions :")
Fraudulent_Transactions.Amount.describe()

Decription of the Amount feature of Fraud transactions :


count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

#### <span style = 'color:Orange'> Compare the mean of both the categories of the transactions </span>
---

In [14]:
creditCard_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9       V10       V11       V12  \
Class                                                                         
0      0.002419  0.009637 -0.000987  0.004467  0.009824 -0.006576  0.010832   
1     -1.397737 -5.568731  0.570636 -2.581123 -5.676883  3.800173 -6.259393   

            V13       V14       V15       V16       V17       V18       V19  \
Class                                                                         
0      0.000189  0.012064  0.000161  0.007164  0.011535  0.003887 -0.001178   
1     -0.109334 -6.971723 -0.092929 -4.139946 -6.665836 -2.246308  0.680659   

            V20       V21       V22       V23       V24       V25       V26  \
Class                                                                         
0     -0.000644 -0.001235 -0.000024  0.000070  0.000182 -0.000072 -0.000089   
1      0.372319  0.713588  0.014049 -0.040308 -0.105130  0.041449  0.051648   

            V27       V28      Amount  
Class                                  
0     -0.000295 -0.000131   88.291022  
1      0.170575  0.075667  122.211321

   * **There is a huge difference in the mean of Legitimate and Fradulant transactions**

#### <span style = 'color:Orange'> Under-Sampling </span>
---

* Since the data contains more than 90% of the legit transactions, it is necessary to build a sample dataset that contains similar distribution of Legit transactions and Fraudulent Transactions.


* Extract 492 random legit transactions and append it to the fraudulent transactions.


* This makes the distribution uniform and the makes it a better dataset for the predictive model.

In [15]:
legitSamples = legitimate_Transactions.sample(n=492)

#### <span style= 'color:Orange'> Uniform Data: Concatinate the Fraud_Transactions and legit_Sample in order to make the distribution of the dataset uniform </span>
---

In [16]:
creditCardData = pd.concat([legitSamples, Fraudulent_Transactions], axis = 0)

#### <span style = 'color:Orange'> Initial rows of the Normalized Data </span>
---

In [18]:
print(f"Initial rows of the Normalised Data:")
creditCardData.head()

Initial rows of the Normalised Data:


Time        V1        V2        V3        V4        V5        V6  \
100981   67687.0  1.277616  0.008527 -1.154306  0.258949  2.193753  3.604186   
17135    28472.0 -1.310665  1.625598  2.148295  2.709668 -0.425643  1.544187   
165197  117266.0  1.973168 -1.255472 -1.747534 -1.084718 -0.358644 -0.385973   
263801  161119.0  2.156286 -0.709245 -1.380799 -0.515675 -0.313343 -0.590552   
166075  117844.0 -1.353424  0.946465  1.726225 -0.660626 -0.429024 -0.342764   

              V7        V8        V9       V10       V11       V12       V13  \
100981 -0.448926  0.851078  0.173330 -0.014826 -0.542440  0.235824 -0.224138   
17135  -0.613314  0.754647 -0.960549  0.869973  0.709577  1.039479  1.012540   
165197 -0.428862 -0.117105 -0.583974  1.011355  0.465711 -0.697996 -0.843387   
263801 -0.299292 -0.218346 -0.416949  0.860852 -1.333736 -0.257078 -0.038958   
166075  0.018457  0.395626  0.047736 -1.018436 -1.259037  0.210871  0.447877   

             V14       V15       V16       V17       V18       V19       V20  \
100981  0.189807 -0.152090 -0.102275 -0.511052 -0.077844  0.365934 -0.053851   
17135  -0.230477 -0.092734 -0.115263  0.220074  0.222998  1.034692  0.003763   
165197  0.400527 -0.127601  1.340217 -0.240743 -0.423289  0.836719  0.213462   
263801  0.183189  0.020243 -1.459221 -0.126234  0.648643 -0.456205 -0.557564   
166075 -0.198422  0.011183  0.384984 -0.382131  0.142788 -0.258217 -0.181569   

             V21       V22       V23       V24       V25       V26       V27  \
100981 -0.187916 -0.476908 -0.120082  1.009497  0.895922 -0.325840  0.026738   
17135   0.210873  0.427363 -0.085391 -0.269936 -0.371199  0.153022 -0.624072   
165197  0.528497  1.133849 -0.144763  0.318023  0.154309  0.039542 -0.076255   
263801 -0.670574 -1.454911  0.406407  0.423226 -0.399927  0.295342 -0.067239   
166075  0.231828  0.555367 -0.475748  0.051700  0.531765  0.667120 -0.372346   

             V28  Amount  Class  
100981  0.011169    1.00      0  
17135   0.042972    0.00      0  
165197 -0.052210  141.65      0  
263801 -0.049198   30.00      0  
166075 -0.027597   11.50      0

#### <span style = 'color:Orange'> Final rows of the Normalized Data </span>
---

In [19]:
print(f"Initial rows of the Normalised Data:")
creditCardData.tail()

Initial rows of the Normalised Data:


Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9       V10       V11       V12       V13  \
279863 -0.882850  0.697211 -2.064945 -5.587794  2.115795 -5.417424 -1.235123   
280143 -1.413170  0.248525 -1.127396 -3.232153  2.858466 -3.096915 -0.792532   
280149 -2.234739  1.210158 -0.652250 -3.463891  1.794969 -2.775022 -0.418950   
281144 -2.208002  1.058733 -1.632333 -5.245984  1.933520 -5.030465 -1.127455   
281674  0.223050 -0.068384  0.577829 -0.888722  0.491140  0.728903  0.380428   

             V14       V15       V16       V17       V18       V19       V20  \
279863 -6.665177  0.401701 -2.897825 -4.570529 -1.315147  0.391167  1.252967   
280143 -5.210141 -0.613803 -2.155297 -3.267116 -0.688505  0.737657  0.226138   
280149 -4.057162 -0.712616 -1.603015 -5.035326 -0.507000  0.266272  0.247968   
281144 -6.416628  0.141237 -2.549498 -4.614717 -1.478138 -0.035480  0.306271   
281674 -1.948883 -0.832498  0.519436  0.903562  1.197315  0.593509 -0.017652   

             V21       V22       V23       V24       V25       V26       V27  \
279863  0.778584 -0.319189  0.639419 -0.294885  0.537503  0.788395  0.292680   
280143  0.370612  0.028234 -0.145640 -0.081049  0.521875  0.739467  0.389152   
280149  0.751826  0.834108  0.190944  0.032070 -0.739695  0.471111  0.385107   
281144  0.583276 -0.269209 -0.456108 -0.183659 -0.328168  0.606116  0.884876   
281674 -0.164350 -0.295135 -0.072173 -0.450261  0.313267 -0.289617  0.002988   

             V28  Amount  Class  
279863  0.147968  390.00      1  
280143  0.186637    0.76      1  
280149  0.194361   77.89      1  
281144 -0.253700  245.00      1  
281674 -0.015309   42.53      1

#### <span style= 'color:Orange'> Number of Legimate Transactions and Fradualent Transactions of the new dataset </span>
---

In [20]:
creditCardData.Class.value_counts()

0    492
1    492
Name: Class, dtype: int64

  * <span style = 'color:Red'> **Now the data is uniformly distributed where there are 492 Legitimate and Fraudulant Transactions.** </span>

#### <span style= 'color:Orange'> Compare the average of both categories of transactions.</span>
---

In [21]:
creditCardData.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94479.148374  0.025021  0.070664  0.055870 -0.014002 -0.083050   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9       V10       V11       V12  \
Class                                                                         
0     -0.001518 -0.001377 -0.092830 -0.043472  0.011638 -0.026260  0.007893   
1     -1.397737 -5.568731  0.570636 -2.581123 -5.676883  3.800173 -6.259393   

            V13       V14       V15       V16       V17       V18       V19  \
Class                                                                         
0      0.038563  0.039110 -0.146146  0.015148  0.000391  0.014883 -0.020545   
1     -0.109334 -6.971723 -0.092929 -4.139946 -6.665836 -2.246308  0.680659   

            V20       V21       V22       V23       V24       V25       V26  \
Class                                                                         
0      0.028949 -0.079060  0.023320  0.010797  0.054728 -0.029514  0.012471   
1      0.372319  0.713588  0.014049 -0.040308 -0.105130  0.041449  0.051648   

            V27       V28      Amount  
Class                                  
0     -0.029081 -0.008941   92.370325  
1      0.170575  0.075667  122.211321

---

## <span style = 'color:Indigo'> SEGREGATE THE DATA INTO DEPENDENT AND INDEPENDET FEATURES </span>
---

In [22]:
X = creditCardData.drop(columns = 'Class', axis = 1)
y = creditCardData['Class']

---

## <span style = 'color:Indigo'> SPLIT THE DATA INTO TRAINING AND TESTING DATASET </span>
---

In [31]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, stratify = y, random_state = 2)

  * <span style = 'color:Red'>**Startify is used to keep the proportion of y values through the training and test sets.**</span>

---

## <span style = 'color:Indigo'> BUILDING THE MODEL </span>
---

### <span style = 'color:Orange'> The Logistics Regression Model </span>
---

   * It is the a machine learning model that comes under **Supervised Learning Technique.**
   
   
   * It is used to predict the categorical dependent variables using a given set of independent variables.

In [33]:
model = LogisticRegression()

### <span style= 'color:Orange'> Training the Logistic Regression Model using the Training Dataset </span>
---

In [37]:
model.fit(X_train, y_train)

LogisticRegression()

### <span style= 'color:Orange'> Model Evaluation. </span>
---

  * Accuracy Score --> Training Data

In [41]:
X_train_prediction = model.predict(X_train)

print(f"The accuracy of the model with training data : {round((accuracy_score(X_train_prediction, y_train) * 100), 2)}%")

The accuracy of the model with training data : 94.54%


  * Accuracy Score --> Testing Data

In [43]:
X_test_Prediction = model.predict(X_test)

print(f"The accuracy of the model with testing data : {round((accuracy_score(X_test_Prediction, y_test) * 100), 2)}%")

The accuracy of the model with testing data : 92.39%


---